Panic in the TestInfo

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
 --- FAIL: TestInfo (56.12s)
    domain_test.go:89: 
        	Error Trace:	domain_test.go:89
        	Error:      	Received unexpected error:
        	            	context deadline exceeded
        	            	github.com/pingcap/errors.AddStack
        	            		/home/prow/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/errors.go:174
        	            	github.com/pingcap/errors.Trace
        	            		/home/prow/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/juju_adaptor.go:15
        	            	github.com/pingcap/tidb/ddl/util.PutKVToEtcd
        	            		/go/tidb/ddl/util/syncer.go:156
        	            	github.com/pingcap/tidb/ddl/util.(*schemaVersionSyncer).Init
        	            		/go/tidb/ddl/util/syncer.go:185
        	            	github.com/pingcap/tidb/domain.(*Domain).Init
        	            		/go/tidb/domain/domain.go:819
        	            	github.com/pingcap/tidb/domain.TestInfo
        	            		/go/tidb/domain/domain_test.go:89
        	            	testing.tRunner
        	            		/usr/local/go/src/testing/testing.go:1259
        	            	runtime.goexit
        	            		/usr/local/go/src/runtime/asm_amd64.s:1581
        	Test:       	TestInfo
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x26fd20f]
goroutine 253 [running]:
testing.tRunner.func1.2({0x38336e0, 0x5b09730})
	/usr/local/go/src/testing/testing.go:1209 +0x24e
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1212 +0x218
panic({0x38336e0, 0x5b09730})
	/usr/local/go/src/runtime/panic.go:1038 +0x215
github.com/pingcap/tidb/ddl.(*ddl).close(0xc00481a180)
	/go/tidb/ddl/ddl.go:421 +0x6f
github.com/pingcap/tidb/ddl.(*ddl).Stop(0xc00481a180)
	/go/tidb/ddl/ddl.go:346 +0x85
github.com/pingcap/tidb/domain.(*Domain).Close(0xc00481c140)
	/go/tidb/domain/domain.go:695 +0x71
github.com/pingcap/tidb/domain.TestInfo.func1()
	/go/tidb/domain/domain_test.go:71 +0x38
runtime.Goexit()
	/usr/local/go/src/runtime/panic.go:642 +0x187
testing.(*common).FailNow(0xc000523040)
	/usr/local/go/src/testing/testing.go:756 +0x45
github.com/stretchr/testify/require.NoError({0x41ceed0, 0xc000523040}, {0x41b2f00, 0xc00494b920}, {0x0, 0x0, 0x0})
	/home/prow/go/pkg/mod/github.com/stretchr/testify@v1.7.0/require/require.go:1234 +0xb4
github.com/pingcap/tidb/domain.TestInfo(0xc000523040)
	/go/tidb/domain/domain_test.go:89 +0x5a8
testing.tRunner(0xc000523040, 0x3d85720)
	/usr/local/go/src/testing/testing.go:1259 +0x102
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1306 +0x35a
FAIL	github.com/pingcap/tidb/domain	56.179s 
```
<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

### 3. What did you see instead (Required)

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->

--- a/domain/domain.go
+++ b/domain/domain.go
@@ -704,7 +704,9 @@ func (do *Domain) Close() {
 	}
 
 	do.slowQuery.Close()
-	do.cancel()
+	if do.cancel != nil {
+		do.cancel()
+	}
 	do.wg.Wait()
 	do.sysSessionPool.Close()
 	variable.UnregisterStatistics(do.bindHandle)

